tests/baselines/reference/narrowingDestructuring.js

```diff
@@ -0,0 +1,89 @@
+//// [narrowingDestructuring.ts]
+type X = { kind: "a", a: string } | { kind: "b", b: string }
+
+function func<T extends X>(value: T) {
+    if (value.kind === "a") {
+        value.a;
+        const { a } = value;
+    } else {
+        value.b;
+        const { b } = value;
+    }
+}
+
+type Z = { kind: "f", f: { a: number, b: string, c: number } }
+    | { kind: "g", g: { a: string, b: number, c: string }};
+
+function func2<T extends Z>(value: T) {
+    if (value.kind === "f") {
+        const { f: f1 } = value;
+        const { f: { a, ...spread } } = value;
```

Comment:
I'm not sure if this case is any more weird than this whole thing already is, but... `...spread` here is not going to have a generic type, since we first get the non-generic type of `value`, then find out the non-generic type of property `f` on that type, and that's the type we use to compute the type of `spread`.